controller/controller_server/exception_handler.py

```diff
@@ -0,0 +1,39 @@
+from decorator import decorator
+
+import grpc
+
+import controller.array_action.errors as controller_errors
+from controller.common.csi_logger import get_stdout_logger
+from controller.controller_server.errors import ValidationException
+
+logger = get_stdout_logger()
+
+status_codes_by_exception = {
+    ValidationException: grpc.StatusCode.INVALID_ARGUMENT,
+    controller_errors.IllegalObjectName: grpc.StatusCode.INVALID_ARGUMENT,
+    controller_errors.SpaceEfficiencyNotSupported: grpc.StatusCode.INVALID_ARGUMENT,
+    controller_errors.PoolDoesNotExist: grpc.StatusCode.INVALID_ARGUMENT,
+    controller_errors.PoolDoesNotMatchCapabilities: grpc.StatusCode.INVALID_ARGUMENT,
+    controller_errors.PoolParameterIsMissing: grpc.StatusCode.INVALID_ARGUMENT,
+    controller_errors.ExpectedSnapshotButFoundVolumeError: grpc.StatusCode.INVALID_ARGUMENT,
+    controller_errors.ObjectNotFoundError: grpc.StatusCode.NOT_FOUND,
+    controller_errors.PermissionDeniedError: grpc.StatusCode.PERMISSION_DENIED,
+    controller_errors.VolumeAlreadyExists: grpc.StatusCode.ALREADY_EXISTS,
+    controller_errors.NotEnoughSpaceInPool: grpc.StatusCode.RESOURCE_EXHAUSTED
+}
+
+
+def handle_common_exceptions(response_type):
+    @decorator
+    def decorated_handle_common_exceptions(controller_method, servicer, request, context):
+        try:
+            return controller_method(servicer, request, context)
+        except Exception as ex:
+            if type(ex) not in status_codes_by_exception.keys():
+                raise
```

Comment:
shouldn't we just return with `grpc.StatusCode.INTERNAL`?

Reply:
The **common** handler shouldn't be familiar with other (uncommon) exceptions and their status codes.
It should transfer the responsibility to the original function.

Reply:
there is no original function, this is our last line of handling.
the way i see it, there are 3 kinds of exceptions:

1. method-specific
2. common
3. other

and they should be handled by:

1. the specific method
2. the decorator
3. the decorator also

since the handling logic of "other" exceptions is also common.

Reply:
Then it's not a "common exceptions handler", but "not method-specific exceptions". I'm not a fan of this design.
How can I know which exception is a method-specific exception?
Some exceptions are used only in one method, but they can be used in other future functions as well.

Reply:
the point here is to move common logic, which is *mostly* about common exceptions.
if the exception in question appears in only one method, or has different handlings, then it's specific.
when new usage adds, its location should be determined by said reasoning.

Reply:
The original point here is removing some try/except clauses and a linter warning.
The point of CSI-2778 is breaking the god object and reduce the code length by breaking the file into a couple of files wrap some code with functions.
I think we should discuss how we can improve the class and redesign it in CSI-2778, but If you want, we can talk about it tomorrow.

Reply:
both of our suggestions serve the original point you described, so I'm not sure how this helps.
I agree that this decorator can easily solve both CSI-2778 and CSI-1358, and I don't see why not

Reply:
Done